### PR TITLE
refactor(azure): resources migrations batch3

### PR DESCRIPTION
**Pull request by sinabakh**

### CODE REVIEW

**Comment by sinabakh** [comment]
@aliscott Ready for review

File: internal/resources/azure/linux_virtual_machine.go [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+	purchaseOptionLabel := "pay as you go"
+
+	productNameRe := "/Virtual Machines .* Series$/"
+	if strings.HasPrefix(instanceType, "Basic_") {
+		productNameRe = "/Virtual Machines .* Series Basic$/"
+	} else if !strings.HasPrefix(instanceType, "Standard_") {
+		instanceType = fmt.Sprintf("Standard_%s", instanceType)
+	}
+
+	qty := decimal.NewFromFloat(730)
```
Comment (aliscott) [inline internal/resources/azure/linux_virtual_machine.go:76]: We could use `schema.HourToMonthUnitMultiplier` here.

Comment (sinabakh) [reply internal/resources/azure/linux_virtual_machine.go]: Nice! There were a few other instances in azure that I also fixed.

File: internal/resources/azure/kubernetes_cluster_node_pool.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+			subResources = append(subResources, osDisk)
+			schema.MultiplyQuantities(subResources[0], nodeCount)
+			mainResource.SubResources = subResources
+		}
+	}
+
+	return mainResource
+}
+
+func aksOSDiskSubResource(region string, diskSize int) *schema.Resource {
```
[the remark this answers is not in the document]
Comment (sinabakh) [reply internal/resources/azure/kubernetes_cluster_node_pool.go:73]: This method is used by `aksClusterNodePool` which is used by many resources so we can't make it a method of struct.

File: internal/resources/azure/application_gateway.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+func (r *ApplicationGateway) v2CostComponents(skuNameParts []string) []*schema.CostComponent {
+	var monthlyCapacityUnits *decimal.Decimal
+	costComponents := make([]*schema.CostComponent, 0)
+
+	if r.MonthlyV2CapacityUnits != nil {
+		monthlyCapacityUnits = decimalPtr(decimal.NewFromInt(*r.MonthlyV2CapacityUnits))
+	}
+	var tier string
+	if strings.ToLower(skuNameParts[0]) == "standard" {
```
[the remark this answers is not in the document]
Comment (sinabakh) [reply internal/resources/azure/application_gateway.go]: Nice catch! fixed

File: internal/providers/terraform/azure/registry.go [hunk 4]
```diff
@@ -7,7 +7,7 @@ var ResourceRegistry []*schema.RegistryItem = []*schema.RegistryItem{
 	getActiveDirectoryDomainServiceRegistryItem(),
 	getActiveDirectoryDomainServiceReplicaSetRegistryItem(),
 	getAPIManagementRegistryItem(),
-	GetAzureRMApplicationGatewayRegistryItem(),
+	getAzureApplicationGatewayRegistryItem(),
```
[the remark this answers is not in the document]
Comment (sinabakh) [reply internal/providers/terraform/azure/registry.go]: I missed it. Fixed.

**Comment by sinabakh** [comment]
@aliscott Your comments are applied.

File: internal/providers/terraform/azure/registry.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (aliscott) [reply internal/providers/terraform/azure/registry.go]: There's still a lot of the `getAzure...` functions in this file. I think we should change them all to just be `get...`.

[the remark this answers is not in the document]
Comment (sinabakh) [reply internal/providers/terraform/azure/registry.go]: I missed what you meant in the comment. All migrated resources were fixed but the resources that are not migrated are not touched yet.

[the remark this answers is not in the document]
Comment (aliscott) [reply internal/providers/terraform/azure/registry.go]: Thanks!

**Review by aliscott — COMMENTED** [review]
@sinabakh this looks almost good to go but I'm getting a panic when running the tests. (Unfortunately we don't run Azure tests for forks in GitHub Actions since they need creds).

```
/usr/local/bin/go test -timeout 60s -run ^TestAzureRMWindowsVirtualMachineScaleSetGoldenFile$ github.com/infracost/infracost/internal/providers/terraform/azure -v
=== RUN   TestAzureRMWindowsVirtualMachineScaleSetGoldenFile
=== PAUSE TestAzureRMWindowsVirtualMachineScaleSetGoldenFile
=== CONT  TestAzureRMWindowsVirtualMachineScaleSetGoldenFile
=== RUN   TestAzureRMWindowsVirtualMachineScaleSetGoldenFile/HCL
--- FAIL: TestAzureRMWindowsVirtualMachineScaleSetGoldenFile (0.02s)
    --- FAIL: TestAzureRMWindowsVirtualMachineScaleSetGoldenFile/HCL (0.02s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1646f63]

goroutine 51 [running]:
testing.tRunner.func1.2({0x22645a0, 0x335a020})
	/usr/local/Cellar/go/1.18.4/libexec/src/testing/testing.go:1389 +0x24e
testing.tRunner.func1()
	/usr/local/Cellar/go/1.18.4/libexec/src/testing/testing.go:1392 +0x39f
panic({0x22645a0, 0x335a020})
	/usr/local/Cellar/go/1.18.4/libexec/src/runtime/panic.go:838 +0x207
github.com/infracost/infracost/internal/resources/azure.(*WindowsVirtualMachineScaleSet).BuildResource(0xc0008b02a0)
	/Users/ali/code/infracost/internal/resources/azure/windows_virtual_machine_scale_set.go:50 +0x163
github.com/infracost/infracost/internal/providers/terraform/azure.NewWindowsVirtualMachineScaleSet(0xc0000d2840, 0xc000851f80?)
	/Users/ali/code/infracost/internal/providers/terraform/azure/windows_virtual_machine_scale_set.go:36 +0x1274
github.com/infracost/infracost/internal/providers/terraform.(*Parser).createResource(0xc0000910c0, 0xc0000d2840, 0xc00000e0d8)
	/Users/ali/code/infracost/internal/providers/terraform/parser.go:57 +0x12b
github.com/infracost/infracost/internal/providers/terraform.(*Parser).parseJSONResources(_, _, {_, _, _}, _, {0x5, {0xc000867400, 0x20ac}, {0x0, ...}, ...}, ...)
	/Users/ali/code/infracost/internal/providers/terraform/parser.go:105 +0x5ef
github.com/infracost/infracost/internal/providers/terraform.(*Parser).parseJSON(0xc0000910c0, {0xc00086e000, 0x20ac, 0x4148}, 0xc000011570?)
	/Users/ali/code/infracost/internal/providers/terraform/parser.go:145 +0x648
github.com/infracost/infracost/internal/providers/terraform.(*HCLProvider).parseResources(0xc000150d90, {0xc0000ca200?, 0x0?}, {0xc00086e000, 0x20ac, 0x4148}, 0x0?)
	/Users/ali/code/infracost/internal/providers/terraform/hcl_provider.go:192 +0x166
github.com/infracost/infracost/internal/providers/terraform.(*HCLProvider).LoadResources(0xc0005809c0?, 0xc0003c07e0?)
	/Users/ali/code/infracost/internal/providers/terraform/hcl_provider.go:170 +0x105
github.com/infracost/infracost/internal/providers/terraform/tftest.loadResources(0xc0005809c0, {0x2460b96, 0x3}, {{0xc0007c7a90?, 0x0?, 0x0?}}, 0xc0003c07e0, 0xc000090120?)
	/Users/ali/code/infracost/internal/providers/terraform/tftest/tftest.go:308 +0x4ad
github.com/infracost/infracost/internal/providers/terraform/tftest.goldenFileResourceTestWithOpts(0xc0005809c0, {0x2460b96, 0x3}, {0x24bd9a5, 0x26}, 0xc0007b6018)
	/Users/ali/code/infracost/internal/providers/terraform/tftest/tftest.go:255 +0x3d4
github.com/infracost/infracost/internal/providers/terraform/tftest.GoldenFileResourceTestsWithOpts.func1(0xc000580680?)
	/Users/ali/code/infracost/internal/providers/terraform/tftest/tftest.go:203 +0x31
testing.tRunner(0xc0005809c0, 0xc0001be0a0)
	/usr/local/Cellar/go/1.18.4/libexec/src/testing/testing.go:1439 +0x102
created by testing.(*T).Run
	/usr/local/Cellar/go/1.18.4/libexec/src/testing/testing.go:1486 +0x35f
FAIL	github.com/infracost/infracost/internal/providers/terraform/azure	6.095s
FAIL
```

**Comment by sinabakh** [comment]
@aliscott I'll take care of it.

**Comment by aliscott** [comment]
@sinabakh have you had a chance to look at the panic at all?

**Comment by sinabakh** [comment]
@aliscott I spent a few hours on it but it's not done.
As you might know, the internet has been practically disconnected here for 2 weeks now and proxies are also slow so it takes a lot to run some terraform tests.

**Comment by sinabakh** [comment]
It's unfortunate but I'm not sure when I would be able to finish this PR so it would be great if anyone could finish it.

**Pull request closed without merging**